### PR TITLE
[TACHYON-336] Parallel Vagrant Deployment

### PR DESCRIPTION
**Pull request by uronce-cc**
https://tachyon.atlassian.net/browse/TACHYON-336

## Summary

Currently, deploy/vagrant deploys machine one by one, when deploying a large cluster, this is not efficient.

So, this PR parallelizes the process! 

Other things happening in the PR include a colorful bash prompt telling you how you can access Tachyon's web console, a script for testing 
deployment time on aws in different scale, skip licence check and style check for efficiency of mvn build.

## Analysis

There are some time consuming steps in the deployment:

- install software, some software are needed by each machine to run hadoop and tachyon like java, but some are just needed by machines to do 
specific job like maven for building tachyon jar from source, so there is no need for each machine to install maven
- downloads, if you are using virtualbox on your desktop, there is no need to download the same contents from the internet in each machine
- build tachyon, the maven build process needs to download a lot of jars, xmls, there is no need for each machien to build their own jar, 
all they need is tachyon's binary distribution
- TachyonMaster has different jobs to do during deployment comparing to TachyonWorkers, so vagrant's --parallel option during provision doesn't work
- Some providers of Vagrant don't support parallel provision, like virtualbox

## Solution

This PR refactors vagrant's provision, replacing shell provision with Ansible(http://docs.ansible.com/), ansible takes care of parallel provision. 
so, the downside of this PR is you need to install ansible and vagrant, but ansible is just a python package, it's very easy to install(http://docs.ansible.com/intro_installation.html)

The above problems in Analysis can be solved:

- For software needed by all machines, they are installed in each machine sequentially, but in parallel among machines
- For software needed by specific machines, they are only installed when needed
- When using virtualbox provider, all downloads(including maven repositories) are shared through /vagrant/shared, so you only need to download once, and 
all the downloads are saved on your host's deploy/vagrant/shared folder. 
- Only TachyonMaster will build tachyon, TachyonWorkers will rsync what they need(like the jar, bin directory...)from TachyonMaster, again, in parallel!
- When you deploy a git repo, only TachyonMaster will clone the repo.
- When you deploy a repo on your host, If you use virtualbox, the repo directory will be mounted to each machine with vagrant's synced_folder feature.
If you use other providers, like aws, the host repo will be rsynced to remote TachyonMaster. Reason behinds the different handling between virtualbox
and other providers here is that synced_folder feature on host should be more efficient than rsync because it should be implemented through virtualbox's ability to mount host directories, I think this is implemented through kernel module instead of network transfer like rsync. 

## Test

Here are the tests I've done, I didn't test the original codes on aws, because when I test on host, I have known what I will feel if I try aws.

**Host Environment**
- OSX     10.10.2
- Vagrant 1.7.2
- Ansible 1.9.0.1

First time I ./run_vb.sh, it is slow because a lot of downloading is happending, all downloads are saved on hosts' vagrant/shared directory.
The second time is much faster, a 3 node cluster only takes around 6min to setup. 

**AWS Environment**
- Region: us-west-2
- Availability_Zone: us-west-2a
- AMI: HVM RHEL7 ami-4dbf9e7d

### Time taken for different cluster scale

time is measured by `time` command on host, so it includes the communication time between host and AWS which may have impact on the tests because I didn't run them all in one place, also, some tests are run with Ansible's verbose mode which causes more network transfer of logs, some are using local repo, some are using github repo, these will vary test time. In summary, the time increase is not linear to the increase of number of nodes.

**Instance Type: t2.small**
<table class="table">
<tr>
  <th>local/github</th><th>verbose_mode</th><th>number_of_nodes</th><th>time_in_minutes</th>
</tr>
<tr>
  <td>local</td><td>Y</td><td>2</td><td>10m50.889s</td>
</tr>
<tr>
  <td>local</td><td>Y</td><td>3</td><td>10m32.000s</td>
</tr>
<tr>
  <td>github</td><td>Y</td><td>5</td><td>10m22.470s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>5</td><td>11m50.102s</td>
</tr>
<tr>
  <td>local</td><td>Y</td><td>10</td><td>15m59.987s</td>
</tr>
</table>
**InstanceType: r3.xlarge**
<table class="table">
<tr>
  <th>local/github</th><th>verbose_mode</th><th>number_of_nodes</th><th>time_in_minutes</th>
</tr>
<tr>
  <td>github</td><td>N</td><td>5</td><td>11m55.908s</td>
</tr>
</table>

### CODE REVIEW

**Comment by uronce-cc** [comment]
@calvinjia @rootfs @haoyuan

**Comment by calvinjia** [comment]
@uronce-cc 
Thanks for the patch! For the timing results, it looks like verbose mode was slower than non-verbose (unexpected?). Also, it seems like deployment time was constant up to 5 nodes, what causes the longer time in 10 node clusters? Is it just a long tail being more evident since there are more machines?

Also, is it absolutely necessary to replace shell provisioning with ansible? What are the features provided? It would be best to reduce the number of things the user needs to set up to run these scripts.

File: deploy/vagrant/tachyon_version.yml [hunk 1]
```diff
@@ -4,4 +4,4 @@ Type: Local
 # github repo and commit hash 
 Github: 
   Repo: https://github.com/amplab/tachyon
-  Hash: eaeaccd19832494c5f0b2a792888a5eaacd7bcf8
+  Hash: 74f2968d99a9bb5f9ed973c1a226eee25d33ba65
```
Comment (calvinjia) [inline deploy/vagrant/tachyon_version.yml]: Not sure if it makes sense to keep updating this hash, maybe just leave it as something the user needs to update (since it very rarely will be what they want).

File: deploy/vagrant/ufs/hadoop2/config_tachyon.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#!/bin/bash
+
+namenode=$(tail -n1 /tachyon/conf/workers)
```
Comment (calvinjia) [inline deploy/vagrant/ufs/hadoop2/config_tachyon.sh]: Why is the formatting different from the command in the hadoop1 script? Also it would be good to not rely on confs outside of the deploy directory, but that can be dealt with in a separate PR.

File: deploy/vagrant/tachyon_version.yml (hunk 1 above)
Comment (uronce-cc) [inline deploy/vagrant/tachyon_version.yml]: agree. After this change, won't do again.

File: deploy/vagrant/ufs/hadoop2/config_tachyon.sh (hunk 2 above)
Comment (uronce-cc) [inline deploy/vagrant/ufs/hadoop2/config_tachyon.sh]: sometimes I like `xxxx`, sometimes I like $(xxx).   Have formatted to `xxx` in all .sh

**Comment by uronce-cc** [comment]
@calvinjia verbose mode should take more time because ansible will transfer more logs from remote to host.

**Comment by calvinjia** [comment]
@uronce-cc do you know what caused the increase in time for 10 nodes? Since there was almost no increase in time from 2 - 5 nodes.

**Comment by uronce-cc** [comment]
@calvin maybe it's because different network situation, I conducted the experiments in different places. 

Here is the experiments I run today under the same network setting:

**InstanceType: r3.xlarge  Under the same network situation**
<table class="table">
<tr>
  <td>github</td><td>N</td><td>2</td><td>8m26.053s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>3</td><td>9m3.176s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>4</td><td>9m26.444s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>5</td><td>9m39.949s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>6</td><td>10m11.106s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>7</td><td>12m50.160s</td>
</tr>
<tr>
  <td>github</td><td>N</td><td>8</td><td>12m20.764s</td>
</tr>
</table>

I'm wondering about the time increase too, I guess it's because AWS brings machines up slower when machine numbers increase? vagrant sends the request to create machine in AWS sequentially. To support this assumption, I am trying to deploy clusters of different sizes without provisioning, will post the result when it's done.

**Comment by uronce-cc** [comment]
I've tried to deploy to AWS without provisioning, here is the result, time increases with increase of cluster size, but not that much.

**InstanceType: t2.small  without provisioning**
<table class="table">
<tr>
  <td>2</td><td>1m43.245s</td>
</tr>
<tr>
  <td>3</td><td>1m44.075s</td>
</tr>
<tr>
  <td>4</td><td>1m46.354s</td>
</tr>
<tr>
  <td>5</td><td>1m52.674s</td>
</tr>
<tr>
  <td>6</td><td>2m0.560s</td>
</tr>
<tr>
  <td>7</td><td>2m4.090s</td>
</tr>
</table>

File: deploy/vagrant/run_aws.sh [hunk 3]
```diff
@@ -3,15 +3,26 @@ set -e
...
\ No newline at end of file
+vagrant up --provider=aws --parallel --no-provision
+# before provision, we have to ensure all machines are up and ssh usable
+vagrant provision
+
+if [[ "$?" == "0" ]]; then
+    HERE=$(dirname $0)
+    pushd $HERE > /dev/null
+    master=`tail -n 1 $HERE/files/hosts | cut -d' ' -f1`
+    master=`cat .vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory | grep Master | cut -d' ' -f2 | cut -d'=' -f2`
```
Comment (calvinjia) [inline deploy/vagrant/run_aws.sh]: Is one of these lines not necessary?

File: deploy/vagrant/core/config_openstack.rb [hunk 4]
```diff
@@ -28,10 +23,7 @@ def config_os(config, i, total, name, version)
...
     os.floating_ip = "auto"
     os.server_name = TAG + name
-    config.vm.provision "shell", inline: 'sudo sed -i -e "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_config'
-    if i == total # last VM starts tachyon
-      config.vm.provision "shell", path: Post
-      config.vm.provision "shell", path: "core/start_tachyon_cluster.sh"
-    end
   end
+
+  config.vm.provision "shell", inline: 'sudo sed -i -e "s/PasswordAuthentication no/PasswordAuthentication yes/g" /etc/ssh/sshd_config'
```
Comment (calvinjia) [inline deploy/vagrant/core/config_openstack.rb]: Is it too messy to put this into the ansible playbook?

File: deploy/vagrant/provision/start_tachyon_cluster.sh [hunk 5]
```diff
@@ -2,5 +2,5 @@
 
 cd /tachyon/bin
 ./tachyon format
-./tachyon-start.sh all Mount
+./tachyon-start.sh all SudoMount
```
Comment (calvinjia) [inline deploy/vagrant/provision/start_tachyon_cluster.sh]: Why does this need to be sudomount now?

**Comment by calvinjia** [comment]
@uronce-cc Thanks for the additional tests, I think they make more sense than what was observed earlier. I left a few more comments but overall it looks good to me.

File: deploy/vagrant/provision/start_tachyon_cluster.sh (hunk 5 above)
Comment (uronce-cc) [inline deploy/vagrant/provision/start_tachyon_cluster.sh]: when I haven't changed the commit hash in tachyon_version.yml, the corresponding tachyon version has permission issues on my mac when mount without sudo, the jira issue, you remember? 

yes, it should be removed now.

**Comment by uronce-cc** [comment]
@calvinjia comments resolved.

File: deploy/vagrant/run_aws.sh [hunk 6]
```diff
@@ -3,15 +3,25 @@ set -e
...
-vagrant provision
\ No newline at end of file
+vagrant up --provider=aws --parallel --no-provision
+# before provision, we have to ensure all machines are up and ssh usable
+vagrant provision
+
+if [[ "$?" == "0" ]]; then
+    HERE=$(dirname $0)
+    master=`tail -n 1 $HERE/files/hosts | cut -d' ' -f1`
+    master=`cat .vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory | grep Master | cut -d' ' -f2 | cut -d'=' -f2`
```
Comment (calvinjia) [inline deploy/vagrant/run_aws.sh]: Oh I actually meant that `master` is being set twice in succession, is it necessary to do both?

Comment (uronce-cc) [inline deploy/vagrant/run_aws.sh]: so ashamed didn't discover.  resolved.

**Comment by calvinjia** [comment]
+1 LGTM
